### PR TITLE
std: detect stack overflows in TLS destructors on UNIX

### LINKED ISSUE #111272: Stack overflow not caught in Drop for TLS data
Rust attempts to catch stack overflow by installing a guard page at the end of the valid stack range. This guard page is stored in TLS data and is torn down when a thread exits (including the main thread).

However other TLS data may run `drop` after this guard page is torn down. Stack overflows occurring in these `drop` implementations are not detected by Rust. (It may be backstopped by the OS, but this is system dependent.)

To reproduce:

```rust
use std::thread_local;
struct First;
struct Second;

impl Drop for First {
    fn drop(&mut self) {
        // First is materialized in TLS in main().
        // Second is materialized in First's dtor, thereby registering a
        // dtor for Second.
        // This dtor is guaranteed to run after the Thread data (including
        // its guard page) has been unmapped.
        SECOND.with(|_s| {} );
    }
}

impl Drop for Second {
    fn drop(&mut self) {
        // Trigger a stack overflow.
        recurse(0);
    }
}

thread_local! {
    static FIRST: First = First;
}

thread_local! {
    static SECOND: Second = Second;
}

// Triggers a stack overflow.
fn recurse(count: usize) {
    if count < usize::MAX {
        recurse(count + 1);
    }
    println!("{} bottles of beer on the wall", count);
}

fn main() {
    FIRST.with(|_f| {});
}
```

This causes a `SIGILL` on macOS and a `SIGSEGV` on Linux. In both cases I confirmed that Rust's stack overflow signal handler is not run.

Reproduced on rust stable and master branch.

### PR DESCRIPTION
**Pull request by joboet**
Fixes #111272.

With #127912 merged, we now have all the infrastructure in place to support stack overflow detection in TLS destructors. This was not possible before because the signal stack was freed in the thread main function, thus a SIGSEGV afterwards would immediately crash. And on platforms without native TLS, the guard page address was stored in an allocation freed in a TLS destructor, so would not be available. #127912 introduced the `local_pointer` macro which allows storing a pointer-sized TLS variable without allocation and the `thread_cleanup` runtime function which is called after all other code managed by the Rust runtime. This PR simply moves the signal stack cleanup to the end of `thread_cleanup` and uses `local_pointer` to store every necessary variable. And so, everything run under the Rust runtime is now properly protected against stack overflows.

### CODE REVIEW

**Comment by Amanieu** [comment]
The logic was a little hard to follow in the code, but after looking at it for a bit it makes sense.

@bors r+

**Comment by jieyouxu** [comment]
I think this [failed in rollup]:

```
testing: silent-thread
testing: loud-thread
testing: silent-thread-tls
--- stderr -------------------------------

thread 'main' panicked at C:\a\rust\rust\tests\ui\runtime\out-of-stack.rs:71:5:
assertion failed: !status.success()
------------------------------------------

failures:
    [ui] tests\ui\runtime\out-of-stack.rs
```

@bors r-

**Comment by ChrisDenton** [comment]
Hmm.. the stack overflow is happening while the thread is already exiting? I'm pretty sure in that case the exiting thread just returns the original exit code rather than changing it. If that's indeed the case (I'll double check when I have a moment) then I don't think there's anything we can reasonably do about it.

**Comment by ChrisDenton** [comment]
Oh right, now that I'm more awake I think I can explain what's going on.

Thread locals drops are run via a TLS callback function. When running the callback, any exceptions thrown (such as the stack overflow exception) will be essential caught and discarded. The callback function will end and further TLS callback functions won't run but neither will the normal exception behaviour.

However, I'm finding it hard to find any official docs about exceptions in TLS callbacks.

**Comment by joboet** [comment]
Mmmh. I guess this was always the case and should be fine... I'll just skip the test.

**Comment by matthiaskrgr** [comment]
@bors r-
failed in https://github.com/rust-lang/rust/pull/136734#issuecomment-2645485952

**Comment by joboet** [comment]
@rustbot ready

I just disabled the signal code check on macOS for now, `abort` doesn't always produce `SIGABRT` when called in a TLS destructor on macOS.

**Comment by jhpratt** [comment]
@bors r-

https://github.com/rust-lang/rust/pull/137506#issuecomment-2677586992

**Comment by clarfonthey** [comment]
r? libs
